worker/gouging.go

```diff
@@ -16,62 +17,90 @@ const keyGougingChecker contextKey = "GougingChecker"
 
 type (
 	GougingChecker interface {
-		Check(rhpv2.HostSettings) GougingResults
+		CheckHS(*rhpv2.HostSettings) GougingResults
+		CheckPT(*rhpv3.HostPriceTable) GougingResults
 	}
 
 	GougingResults struct {
-		downloadErr     error
-		formContractErr error
-		uploadErr       error
+		downloadErr error
+		gougingErr  error
+		uploadErr   error
 	}
 
 	gougingChecker struct {
-		settings    api.GougingSettings
-		minShards   int
-		totalShards int
+		settings   api.GougingSettings
+		redundancy api.RedundancySettings
 	}
 
 	contextKey string
 )
 
-func PerformGougingChecks(ctx context.Context, hs rhpv2.HostSettings) GougingResults {
-	if gc, ok := ctx.Value(keyGougingChecker).(GougingChecker); ok {
-		return gc.Check(hs)
+func PerformGougingChecks(ctx context.Context, hs *rhpv2.HostSettings, pt *rhpv3.HostPriceTable) (results GougingResults) {
+	gc, ok := ctx.Value(keyGougingChecker).(GougingChecker)
+	if !ok {
+		panic("no gouging checker attached to the context") // developer error
 	}
-	panic("no gouging checker attached to the context") // developer error
+
+	results.merge(gc.CheckHS(hs))
+	results.merge(gc.CheckPT(pt))
+	return
 }
 
 func WithGougingChecker(ctx context.Context, gp api.GougingParams) context.Context {
 	return context.WithValue(ctx, keyGougingChecker, gougingChecker{
-		settings:    gp.GougingSettings,
-		minShards:   gp.RedundancySettings.MinShards,
-		totalShards: gp.RedundancySettings.TotalShards,
+		settings:   gp.GougingSettings,
+		redundancy: gp.RedundancySettings,
 	})
 }
 
-func IsGouging(gs api.GougingSettings, hs rhpv2.HostSettings, minShards, totalShards int) (bool, string) {
-	errs := filterErrors(
-		checkDownloadGouging(gs, hs, minShards, totalShards),
-		checkFormContractGouging(gs, hs),
-		checkUploadGouging(gs, hs, minShards, totalShards),
-	)
+func IsGouging(gs api.GougingSettings, rs api.RedundancySettings, hs *rhpv2.HostSettings, pt *rhpv3.HostPriceTable) (bool, string) {
+	var errs []error
+	if hs != nil {
+		errs = append(errs,
+			filterErrors(
+				checkDownloadGouging(gs, rs, hs.BaseRPCPrice, hs.SectorAccessPrice, hs.DownloadBandwidthPrice),
+				checkPriceGougingHS(gs, hs),
+				checkUploadGouging(gs, rs, hs.BaseRPCPrice, hs.SectorAccessPrice, hs.UploadBandwidthPrice),
+			)...,
+		)
+	}
+
+	if pt != nil {
+		errs = append(errs,
+			filterErrors(
+				checkDownloadGouging(gs, rs, pt.InitBaseCost, pt.ReadBaseCost.Add(pt.ReadLengthCost.Mul64(modules.SectorSize)), pt.DownloadBandwidthCost),
+				checkPriceGougingPT(gs, pt),
+				checkUploadGouging(gs, rs, pt.InitBaseCost, pt.WriteBaseCost.Add(pt.WriteLengthCost.Mul64(modules.SectorSize)), pt.UploadBandwidthCost),
+			)...,
+		)
+	}
+
 	if len(errs) == 0 {
 		return false, ""
 	}
+	return true, joinErrors(errs...).Error()
+}
 
-	var reasons []string
-	for _, err := range errs {
-		reasons = append(reasons, err.Error())
+func (gc gougingChecker) CheckHS(hs *rhpv2.HostSettings) (results GougingResults) {
```

Comment:
(nit) we have some duplicate code. `IsGouging` is doing the same as `CheckHS` and `CheckPT`.

Also I do feel a bit uneasy about all the pointers and `if foo != nil { // check }`. Because if we aren't careful we might end up not checking anything due to a dev mistake.

Reply:
I see your point about the `nil` checks, but in this case I don't think it's that much of an issue? It's not like it's pointer magic, there's just a lot of places in our codebase where we have settings and don't care about price tables and vice versa.

Eventually we'll be phasing out host settings anyway right so this code should eventually drop host settings. Most importantly I wanted to avoid having two version for every check (one for HS, one for PT). I removed the duplicate code and added a sanity check.